### PR TITLE
HDDS-7234. Add a common option for DiskBalancer commands

### PR DESCRIPTION
**Pull request by symious**
## What changes were proposed in this pull request?

This ticket is to add a common option for DiskBalancer CLI.

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-7234

## How was this patch tested?

original unit test.

### CODE REVIEW

File: hadoop-hdds/tools/src/main/java/org/apache/hadoop/hdds/scm/cli/datanode/DiskBalancerCommonOptions.java [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+  private List<String> hosts = new ArrayList<>();
+
+  /**
+   * Check the common options of DiskBalancerCommand.
+   * @return if the check passed
+   */
+  public boolean check() {
+    if (hosts.size() == 0 && !allHosts) {
+      System.out.println("Datanode not specified. Please specify " +
+          "\"--allDatanodes\" to start diskBalancer on all datanodes");
```
Comment (dineshchitlangia) [inline hadoop-hdds/tools/src/main/java/org/apache/hadoop/hdds/scm/cli/datanode/DiskBalancerCommonOptions.java:45]:
```suggestion
      System.out.println("Datanode not specified. Please specify at least one datanode or use " +
          "\"--allDatanodes\" to start diskBalancer on all datanodes");
```

File: hadoop-hdds/tools/src/main/java/org/apache/hadoop/hdds/scm/cli/datanode/DiskBalancerCommonOptions.java [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+      return false;
+    }
+    if (hosts.size() != 0 && allHosts) {
+      System.out.println("Confused options. Omit \"-a(--allDatanodes)\" or " +
+          "--hosts.");
```
Comment (dineshchitlangia) [inline hadoop-hdds/tools/src/main/java/org/apache/hadoop/hdds/scm/cli/datanode/DiskBalancerCommonOptions.java:50]:
```suggestion
      System.out.println("Invalid option selection. Use either \"-a(--allDatanodes)\" or " +
          "--hosts.");
```

File: hadoop-hdds/tools/src/main/java/org/apache/hadoop/hdds/scm/cli/datanode/DiskBalancerCommonOptions.java [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Common options for DiskBalancer commands.
+ */
+public class DiskBalancerCommonOptions {
+
+  @CommandLine.Option(names = {"-a", "--allDatanodes"},
+      description = "Run commands on all datanodes.")
```
Comment (dineshchitlangia) [inline hadoop-hdds/tools/src/main/java/org/apache/hadoop/hdds/scm/cli/datanode/DiskBalancerCommonOptions.java:31]: Consider replacing `--allDatanodes` with just `--all`

File: hadoop-hdds/tools/src/main/java/org/apache/hadoop/hdds/scm/cli/datanode/DiskBalancerCommonOptions.java [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+  private boolean allHosts;
+
+  @CommandLine.Option(names = {"--hosts"},
+      description = "Run commands on specific datanodes.")
```
Comment (dineshchitlangia) [inline hadoop-hdds/tools/src/main/java/org/apache/hadoop/hdds/scm/cli/datanode/DiskBalancerCommonOptions.java:35]: To maintain consistency with other options, we should have both short and a long option name -> `-h` and `--hosts`
If `-h` is used for help or conflicts with any existing options, consider using `-d` and `--datanodes`

**Review by dineshchitlangia — CHANGES_REQUESTED** [review]
Overall looks good, suggested minor changes.
Thank you for your contribution.

File: hadoop-hdds/tools/src/main/java/org/apache/hadoop/hdds/scm/cli/datanode/DiskBalancerCommonOptions.java [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+ * Common options for DiskBalancer commands.
+ */
+public class DiskBalancerCommonOptions {
+
+  @CommandLine.Option(names = {"-a", "--all"},
+      description = "Run commands on all datanodes.")
+  private boolean allHosts;
+
+  @CommandLine.Option(names = {"-d", "--datanodes"},
+      description = "Run commands on specific datanodes.")
```
Comment (lokeshj1703) [inline hadoop-hdds/tools/src/main/java/org/apache/hadoop/hdds/scm/cli/datanode/DiskBalancerCommonOptions.java:35]: Could we add some more info on how these datanodes should be specified? Like a comma separated list of datanodes ip or hostname. And we can provide an example as well.

File: hadoop-hdds/tools/src/main/java/org/apache/hadoop/hdds/scm/cli/datanode/DiskBalancerStartSubcommand.java [hunk 6]
```diff
@@ -55,31 +54,23 @@ public class DiskBalancerStartSubcommand extends ScmSubcommand {
...
-      System.out.println("Confused options. Omit \"--allDatanodes\" or " +
-          "Datanodes.");
+    if (!commonOptions.check()) {
       return;
     }
     List<DatanodeAdminError> errors =
         scmClient.startDiskBalancer(threshold, bandwidthInMB, parallelThread,
-            hosts.size() == 0 ? Optional.empty() : Optional.of(hosts));
+            commonOptions.getHosts().size() == 0 ? Optional.empty() :
+                Optional.of(commonOptions.getHosts()));
```
Comment (lokeshj1703) [inline hadoop-hdds/tools/src/main/java/org/apache/hadoop/hdds/scm/cli/datanode/DiskBalancerStartSubcommand.java:69]: Create a function in DiskBalancerCommonOptions for this. We can reuse it in other classes as well.

File: hadoop-hdds/tools/src/main/java/org/apache/hadoop/hdds/scm/cli/datanode/DiskBalancerStartSubcommand.java [hunk 7]
```diff
@@ -55,31 +54,23 @@ public class DiskBalancerStartSubcommand extends ScmSubcommand {
...
 
     System.out.println("Start DiskBalancer on datanode(s):\n" +
-        (allHosts ? "All datanodes" : String.join("\n", hosts)));
+        (commonOptions.isAllHosts() ? "All datanodes" :
+            String.join("\n", commonOptions.getHosts())));
```
Comment (lokeshj1703) [inline hadoop-hdds/tools/src/main/java/org/apache/hadoop/hdds/scm/cli/datanode/DiskBalancerStartSubcommand.java:73]: Create a function in DiskBalancerCommonOptions for this. We can reuse it in other classes as well.

**Review by lokeshj1703 — COMMENTED** [review]
@symious Thanks for working on this! The changes look good to me. I have a few comments inline.

**Comment by symious** [comment]
@dineshchitlangia @lokeshj1703 Updated the PR, please have a look.

**Review by ferhui — APPROVED** [review]

File: hadoop-hdds/tools/src/main/java/org/apache/hadoop/hdds/scm/cli/datanode/DiskBalancerStartSubcommand.java (hunk 6 above)
Comment (lokeshj1703) [reply hadoop-hdds/tools/src/main/java/org/apache/hadoop/hdds/scm/cli/datanode/DiskBalancerStartSubcommand.java:69]: @symious Can you please address this as well? I am +1 on the PR o.w.

**Comment by ferhui** [comment]
@dineshchitlangia @lokeshj1703 Hi, do you have other comments? Plan to merge it tmr.

File: hadoop-hdds/tools/src/main/java/org/apache/hadoop/hdds/scm/cli/datanode/DiskBalancerStartSubcommand.java (hunk 6 above)
Comment (symious) [reply hadoop-hdds/tools/src/main/java/org/apache/hadoop/hdds/scm/cli/datanode/DiskBalancerStartSubcommand.java]: @lokeshj1703 Updated the patch, please have a look.

**Review by lokeshj1703 — APPROVED** [review]
Thanks @symious for updating the PR! The changes look good to me. +1.

**Comment by ferhui** [comment]
@symious Thanks for your contribution! @dineshchitlangia @lokeshj1703 Thanks for your careful review! Merged

**Comment by symious** [comment]
@ferhui @lokeshj1703 @dineshchitlangia Thank you for the review.

**Pull request merged**